2229: Don't move out of drop type

arora-aman opened the pull request:
Fixes #88476

r? @nikomatsakis

nikomatsakis commented:
@rfcbot fcp merge

As discussed in our [last triage meeting], this is an interesting semantic question and a tweak to the closure rules. I've written up the [my conclusion](https://hackmd.io/AC5EJng9SDStdsjG2M9cUQ?view#Niko%E2%80%99s-conclusion) in this hackmd, which also includes a number of other details. I wanted to do an FCP to get ✅ from the @rust-lang/lang team that they've seen this change. There is some urgency here owing to the Rust 2021 edition timeline.

nbdd0121 commented on compiler/rustc_typeck/src/check/upvar.rs:
```diff
@@ -626,7 +633,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 lint::builtin::RUST_2021_INCOMPATIBLE_CLOSURE_CAPTURES,
                 closure_hir_id,
-                closure_head_span,
+                 closure_head_span,
```
I am a bit surprised that rustfmt isn't complaining about this extra space.

arora-aman replied:
I'm surprised `./x.py fmt` made this change

camelid commented:
@arora-aman Not a big deal at all, but was it intentional that you renamed this PR from `2229: Handle drop_defined_for_type during migration` to `2229: Don't move out of drop type 3d7c237`?

arora-aman commented:
> @arora-aman Not a big deal at all, but was it intentional that you renamed this PR from `2229: Handle drop_defined_for_type during migration` to `2229: Don't move out of drop type 3d7c237`?

wwops, the commit didn't need to be there. But yes we changed how we handle the issue -- so I updated the title

camelid commented:
(Also just noting that I think this will need to be backported to beta now that the branch has happened.)

scottmcm commented:
I was convinced in the triage meeting that there are already enough consequences of `Drop` and `Copy` that it's not weird for things to be trait-dependent here.  And since it only comes up in `move` closures, being more aggressive about moving larger things there seems reasonable to me.

@rfcbot reviewed

nikomatsakis commented:
@bors r+ p=1

Rust 2021 related

bors commented:
:pushpin: Commit 153aa71c14d39ee1cb7f989d89b1f2891fdfcb6d has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 153aa71c14d39ee1cb7f989d89b1f2891fdfcb6d -->
<!-- homu: {"type":"Approved","sha":"153aa71c14d39ee1cb7f989d89b1f2891fdfcb6d","approver":"nikomatsakis"} -->

bors commented:
:hourglass: Testing commit 153aa71c14d39ee1cb7f989d89b1f2891fdfcb6d with merge 0d0d2fe182e8980146e25cfcfff0c481892f81c6...
<!-- homu: {"type":"BuildStarted","head_sha":"153aa71c14d39ee1cb7f989d89b1f2891fdfcb6d","merge_sha":"0d0d2fe182e8980146e25cfcfff0c481892f81c6"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3540625289)
Approved by: nikomatsakis
Pushing 0d0d2fe182e8980146e25cfcfff0c481892f81c6 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nikomatsakis","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3540625289"},"merge_sha":"0d0d2fe182e8980146e25cfcfff0c481892f81c6"} -->

rust-timer commented:
Finished benchmarking commit (0d0d2fe182e8980146e25cfcfff0c481892f81c6): [comparison url](https://perf.rust-lang.org/compare.html?start=fdf65053e99e8966f9bd83b5a8491326cb33d638&end=0d0d2fe182e8980146e25cfcfff0c481892f81c6).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

 

@rustbot label: -perf-regression

apiraino commented:
Beta backport accepted as per compiler team [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202021-09-09.20.2354818/near/252633238)

@rustbot label +beta-accepted
